content/courses/Ada_For_The_C_Embedded_Developer/chapters/06_Translation.rst

```diff
@@ -725,9 +725,467 @@ above could also be literally translated to:
+
+The only difference in this case is that the first bit is :ada:`B (1)` instead
+of :ada:`B (0)`.
+
+In C, we would rely on bit-shifting and masking to set that specific bit:
+
+[C]
+
+.. code:: c manual_chop run_button project=Courses.Ada_For_C_Embedded_Dev.Translation.Bitfield_C
+
+    !bitfield.c
+    #include <stdio.h>
+
+    int main(int argc, const char * argv[])
+    {
+        int v = 0;
+
+        v = v | (1 << 2);
+
+        printf("v = %d\n", v);
+    }
+
+We can use this pattern for objects of more complex data types like arrays or
+records. For example:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Translation.Bitfield_Int_Array_Ada
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Int_Array_Bitfield is
+       type Bit_Field is array (Natural range <>) of Boolean with Pack;
+
+       A : array (1 .. 2) of Integer := (others => 0);
+       B : Bit_Field (0 .. A'Size - 1) with Address => A'Address;
+    begin
+       B (2) := True;
+       for I in A'Range loop
+          Put_Line ("A (" & Integer'Image (I)
+                    & ")= " & Integer'Image (A (I)));
+       end loop;
+    end Int_Array_Bitfield;
+
+In the Ada example above, we're using the bit-field to set bit #3 of the first
+element of the array (:ada:`A (1)`). We could set bit #4 of the second element
+by using the size of the data type (in this case, :ada:`Integer'Size`):
+
+[Ada]
+
+.. code-block:: ada
+
+    B (Integer'Size + 3) := True;
+
+In C, we would select the specific array position and, again, rely on
+bit-shifting and masking to set that specific bit:
+
+[C]
+
+.. code:: c manual_chop run_button project=Courses.Ada_For_C_Embedded_Dev.Translation.Bitfield_Int_Array_C
+
+    !bitfield_int_array.c
+    #include <stdio.h>
+
+    int main(int argc, const char * argv[])
+    {
+        int i;
+        int a[2] = {0, 0};
+
+        a[0] = a[0] | (1 << 2);
+
+        for (i = 0; i < 2; i++)
+        {
+            printf("a[%d] = %d\n", i, a[i]);
+        }
+    }
+
+Since we can use this pattern for any arbitrary data type, this allows us to
+easily create a subprogram to serialize data types and, for example, transmit
+complex data structures as a bitstream. For example:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Translation.Bitfield_Serialization_ada
+
+    package Serializer is
+
+       type Bit_Field is array (Natural range <>) of Boolean with Pack;
+
+       procedure Transmit (B : Bit_Field);
+
+    end Serializer;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body Serializer is
+
+       procedure Transmit (B : Bit_Field) is
+
+          procedure Show_Bit (V : Boolean) is
+          begin
+             case V is
+                when False => Put ("0");
+                when True  => Put ("1");
+             end case;
+          end Show_Bit;
+
+       begin
+          Put ("Bits: ");
+          for I in B'Range loop
```

Comment:
Perhaps it's a good place to use "for .. of " instead?

for I of B loop
  Show_Bit (I);
end loop;